pkg/virt-controller/watch/vm.go

```diff
@@ -83,222 +126,433 @@ func (c *VMController) runWorker() {
 }
 
 func (c *VMController) Execute() bool {
-	key, quit := c.queue.Get()
+	key, quit := c.Queue.Get()
 	if quit {
 		return false
 	}
-	defer c.queue.Done(key)
+	defer c.Queue.Done(key)
 	err := c.execute(key.(string))
 
 	if err != nil {
 		log.Log.Reason(err).Infof("reenqueuing VM %v", key)
-		c.queue.AddRateLimited(key)
+		c.Queue.AddRateLimited(key)
 	} else {
 		log.Log.V(4).Infof("processed VM %v", key)
-		c.queue.Forget(key)
+		c.Queue.Forget(key)
 	}
 	return true
 }
 
 func (c *VMController) execute(key string) error {
 
 	// Fetch the latest Vm state from cache
-	obj, exists, err := c.store.GetByKey(key)
+	obj, exists, err := c.vmInformer.GetStore().GetByKey(key)
 
 	if err != nil {
 		return err
 	}
 
-	// Retrieve the VM
-	var vm *kubev1.VirtualMachine
+	// Once all finalizers are removed the vm gets deleted and we can clean all expectations
 	if !exists {
-		namespace, name, err := cache.SplitMetaNamespaceKey(key)
-		if err != nil {
-			return err
-		}
-		vm = kubev1.NewVMReferenceFromNameWithNS(namespace, name)
-	} else {
-		vm = obj.(*kubev1.VirtualMachine)
+		c.podExpectations.DeleteExpectations(key)
+		c.handoverExpectations.DeleteExpectations(key)
+		return nil
 	}
+	vm := obj.(*virtv1.VirtualMachine)
 
 	// If the VM is exists still, don't process the VM until it is fully initialized.
 	// Initialization is handled by the initialization controller and must take place
 	// before the VM is acted upon.
-	if exists && !isVirtualMachineInitialized(vm) {
+	if !isVirtualMachineInitialized(vm) {
 		return nil
 	}
 
 	logger := log.Log.Object(vm)
 
-	if !exists {
-		// Delete VM Pods
-		err := c.vmService.DeleteVMPod(vm)
-		if err != nil {
-			logger.Reason(err).Error("Deleting VM target Pod failed.")
-			return err
-		}
-		logger.Info("Deleting VM target Pod succeeded.")
+	// Get all pods from the namespace
+	pods, err := c.listPodsFromNamespace(vm.Namespace)
+
+	if err != nil {
+		logger.Reason(err).Error("Failed to fetch pods for namespace from cache.")
+		return err
+	}
+
+	// Only consider pods which belong to this vm
+	pods, err = c.filterMatchingPods(vm, pods)
+	if err != nil {
+		return err
+	}
+
+	if len(pods) > 1 {
+		logger.Reason(fmt.Errorf("More than one pod detected")).Error("That should not be possible, will not requeue")
 		return nil
+	} else if len(pods) == 1 && pods[0].Annotations[virtv1.OwnedByAnnotation] == "virt-handler" {
+		// Whenever we see that a matching pod is owned by virt-handler, we can mak handover as fulfilled
+		c.handoverExpectations.CreationObserved(key)
```

Comment:
would it make more sense to place this 'CreationObserved' call in the updatePod() function?

Reply:
Done. Placed it in updatePod and deletePod, since a graceful delete will be "redirected" from updatePod to deletePod.